atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCells.java

```diff
@@ -0,0 +1,251 @@
+        addCell(info, WriteReference.DUMMY, false, 0, -numberDedicatedRows(writes), cells);
+    }
+
+    private void addWrite(PartitionInfo info, WriteInfo write, boolean dedicate, long index, Map<Cell, byte[]> cells) {
+        addCell(info, write.writeRef(), dedicate, index / MAX_CELLS_DEDICATED, index % MAX_CELLS_DEDICATED, cells);
+    }
+
+    private void addCell(PartitionInfo info, WriteReference writeRef, boolean dedicate, long dedicatedRow,
+            long index, Map<Cell, byte[]> cells) {
+        SweepableCellsTable.SweepableCellsRow row = computeRow(info, dedicate, dedicatedRow);
+        SweepableCellsTable.SweepableCellsColumnValue colVal = createColVal(info.timestamp(), index, writeRef);
+        cells.put(SweepQueueUtils.toCell(row, colVal), colVal.persistValue());
+    }
+
+    private long numberDedicatedRows(List<WriteInfo> writes) {
+        return 1 + (writes.size() - 1) / MAX_CELLS_DEDICATED;
+    }
+
+    private static long tsMod(long timestamp) {
+        return timestamp % SweepQueueUtils.TS_FINE_GRANULARITY;
+    }
+
+    private SweepableCellsTable.SweepableCellsColumnValue createColVal(long ts, long index, WriteReference writeRef) {
+        SweepableCellsTable.SweepableCellsColumn col = SweepableCellsTable.SweepableCellsColumn.of(tsMod(ts), index);
+        return SweepableCellsTable.SweepableCellsColumnValue.of(col, writeRef);
+    }
+
+    private void populateWrites(SweepableCellsTable.SweepableCellsRow row, Map.Entry<Cell, Value> entry,
+            Map<CellReference, WriteInfo> writes) {
+        SweepableCellsTable.SweepableCellsColumn col = computeColumn(entry);
+
+        if (isReferenceToDedicatedRows(col)) {
+            addWritesFromDedicated(row, col, writes);
+        } else {
+            addWriteFromValue(getTimestamp(row, col), entry.getValue(), writes);
+        }
+    }
+
+    private boolean isReferenceToDedicatedRows(SweepableCellsTable.SweepableCellsColumn col) {
+        return col.getWriteIndex() < 0;
+    }
+
+    private void addWritesFromDedicated(SweepableCellsTable.SweepableCellsRow row,
+            SweepableCellsTable.SweepableCellsColumn col, Map<CellReference, WriteInfo> writes) {
+        List<byte[]> dedicatedRows = computeDedicatedRows(row, col);
+        RowColumnRangeIterator iterator = getColumnRangeAll(dedicatedRows);
+        iterator.forEachRemaining(entry -> addWriteFromValue(getTimestamp(row, col), entry.getValue(), writes));
+    }
+
+    private void addWriteFromValue(long timestamp, Value value, Map<CellReference, WriteInfo> writes) {
+        WriteReference writeRef = SweepableCellsTable.SweepableCellsColumnValue.hydrateValue(value.getContents());
+        updateLatestForCell(timestamp, writeRef, writes);
+    }
+
+    private void updateLatestForCell(long ts, WriteReference writeRef, Map<CellReference, WriteInfo> writes) {
+        WriteInfo newWrite = WriteInfo.of(writeRef, ts);
+        writes.merge(writeRef.cellReference(), newWrite,
```

Comment:
ok, so we're doing some kind of merge here which means we need to know what the old value was. Is it possible/efficient to do that at a higher level to avoid all the write passing?